File: app/src/main/java/com/zack/enderweather/model/network/NetworkHelper.java
package com.zack.enderweather.model.network;

import com.zack.enderweather.model.bean.HeWeather;
import com.zack.enderweather.util.LogUtil;

import java.io.IOException;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class NetworkHelper {

    private static final String LOG_TAG = "NetworkHelper";

    private static final String URL_HE_WEATHER_API = "https://api.heweather.com/x3/";
    private static final String KEY_HE_WEATHER_API = "b62ec297adac4f64a4aabdc2f86e1ce7";

    /** 异步获取和风天气数据 */
    public void getHeWeatherDataAsync(String cityId, final HeWeatherDataCallback callback) {
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(URL_HE_WEATHER_API)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
        WeatherService service = retrofit.create(WeatherService.class);
        Call<HeWeather> call = service.getHeWeatherData(cityId, KEY_HE_WEATHER_API);
        call.enqueue(new Callback<HeWeather>() {
            @Override
            public void onResponse(Call<HeWeather> call, Response<HeWeather> response) {
                if (response.body() != null) {
                    //解析成功
                    callback.onSuccess(response.body());
                } else {
                    try {
                        LogUtil.e(LOG_TAG, response.errorBody().string());
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                    callback.onFailure(call.request().url().queryParameter("cityid"));
                }
            }

            @Override
            public void onFailure(Call<HeWeather> call, Throwable t) {
                LogUtil.e(LOG_TAG, t.getMessage());
                callback.onFailure(call.request().url().queryParameter("cityid"));
            }
        });
    }

    public interface HeWeatherDataCallback {
        void onSuccess(HeWeather heWeather);
        void onFailure(String cityId);
    }
}

File: app/src/main/java/me/imzack/app/cold/model/network/NetworkHelper.kt
package me.imzack.app.cold.model.network

import com.facebook.stetho.okhttp3.StethoInterceptor
import io.reactivex.Single
import io.reactivex.functions.BiFunction
import me.imzack.app.cold.common.Constant
import me.imzack.app.cold.model.network.bean.HeWeatherApi
import me.imzack.app.cold.model.network.service.HeWeatherApiService
import me.imzack.app.cold.model.network.service.HeWeatherSearchService
import me.imzack.app.cold.util.LogUtil
import me.imzack.app.cold.util.HeWeatherUtil
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory

// 按数据需求组织，不按 API 组织
class NetworkHelper {

    private val heWeatherApiService
        get() = Retrofit.Builder()
                .baseUrl(Constant.HE_WEATHER_API_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .client(getOkHttpClient("HeWeatherApiService"))
                .build()
                .create(HeWeatherApiService::class.java)

    /** 通过城市 id 获取和风天气与空气数据 */
    fun getHeWeatherAndAirData(cityId: String): Single<Pair<HeWeatherApi.Weather, HeWeatherApi.Air>> {
        val service = heWeatherApiService
        val time = HeWeatherUtil.apiSignatureTime
        val sign = HeWeatherUtil.makeApiSignature(cityId, time)
        return Single.zip(
                service.getWeatherData(cityId, Constant.HE_WEATHER_USER_ID, time, sign),
                service.getAirData(cityId, Constant.HE_WEATHER_USER_ID, time, sign),
                // SAM 转换
                BiFunction<HeWeatherApi.Weather, HeWeatherApi.Air, Pair<HeWeatherApi.Weather, HeWeatherApi.Air>> { hwaWeather, hwaAir -> Pair(hwaWeather, hwaAir) }
        )
    }

    /** 通过城市 id 获取和风天气数据 */
    fun getHeWeatherData(cityId: String): Single<HeWeatherApi.Weather> {
        val time = HeWeatherUtil.apiSignatureTime
        return heWeatherApiService.getWeatherData(cityId, Constant.HE_WEATHER_USER_ID, time, HeWeatherUtil.makeApiSignature(cityId, time))
    }

    /** 通过经纬度查询城市 id */
    fun getHeWeatherCityData(longitude: Double, latitude: Double) = Retrofit.Builder()
            .baseUrl(Constant.HE_WEATHER_SEARCH_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .client(getOkHttpClient("HeWeatherSearchService"))
            .build()
            .create(HeWeatherSearchService::class.java)
            .getFindData("$longitude,$latitude", Constant.HE_WEATHER_API_KEY)

    /**
     * 获取一个 OkHttpClient 对象
     * @param tag Logcat 打印的 tag
     */
    private fun getOkHttpClient(tag: String) = OkHttpClient.Builder()
            .addInterceptor(HttpLoggingInterceptor(HttpLoggingInterceptor.Logger { LogUtil.i(it, tag) }).setLevel(HttpLoggingInterceptor.Level.BODY))
            .addNetworkInterceptor(StethoInterceptor())
            .build()
}
